src/ott/solvers/iterative_linear.py

```diff
@@ -0,0 +1,303 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax.experimental import host_callback
+
+from ott import utils
+from ott.math import fixed_point_loop
+from ott.problems.linear import linear_problem
+from ott.solvers import was_solver
+from ott.solvers.linear import sinkhorn, sinkhorn_lr
+
+__all__ = ["IterativeLinearSolver", "IterLinState"]
+
+LinearOutput = Union[sinkhorn.SinkhornOutput, sinkhorn_lr.LRSinkhornOutput]
+
+ProgressCallbackFn_t = Callable[
+    [Tuple[np.ndarray, np.ndarray, np.ndarray, "IterLinState"]], None]
+
+
+class IterLinState(NamedTuple):
+  """State of the IterativeLinearSolver.
+
+  Attributes:
+  costs: Holds the sequence of costs seen through the outer
+  loop of the solver.
+  linear_convergence: Holds the sequence of bool convergence flags of the
+  inner linear solvers.
+  linear_pb: last inner linear_problem
+  linear_state: solution to the linear_pb
+  prob_state: a problem specific state, provided and managed by
+    user-defined functions
+  rngs: random keys passed to the user defined functions at each iteration.
+  errors: Holds sequence of vectors of errors of the Sinkhorn algorithm
+  at each iteration.
+  """
+
+  costs: jnp.ndarray
+  linear_convergence: jnp.ndarray
+  linear_pb: linear_problem.LinearProblem
+  linear_state: LinearOutput
+  prob_state: NamedTuple = None
+  rngs: Optional[jax.random.PRNGKeyArray] = None
+  errors: Optional[jnp.ndarray] = None
+
+  def set(self, **kwargs: Any) -> "IterLinState":
+    """Return a copy of self, possibly with overwrites."""
+    return self._replace(**kwargs)
+
+  def update(  # noqa: D102
+      self,
+      iteration: int,
+      cost: float,
+      linear_pb: linear_problem.LinearProblem,
+      linear_sol: LinearOutput,
+      prob_state: NamedTuple,
+      store_errors: bool,
+  ) -> "IterLinState":
+    costs = self.costs.at[iteration].set(cost)
+    errors = None
+    if store_errors and self.errors is not None:
+      errors = self.errors.at[iteration, :].set(linear_sol.errors)
+    linear_convergence = self.linear_convergence.at[iteration].set(
+        linear_sol.converged
+    )
+    return self.set(
+        linear_pb=linear_pb,
+        linear_state=linear_sol,
+        prob_state=prob_state,
+        costs=costs,
+        linear_convergence=linear_convergence,
+        errors=errors,
+    )
+
+
+@jax.tree_util.register_pytree_node_class
+class IterativeLinearSolver(was_solver.WassersteinSolver):
```

Comment:
@mahegz sorry for the delay

After taking a closer look at the PR and the notebook, I think many of the things in the notebook can be moved from there into the code (esp. the `gcg`). Here's what I propose:
- rather than aiming for a very general solution, I think it will be better to have a good GCG implementation in OTT-JAX + the regularizations you have implemented in the notebook (Laplace + group lasso)
- the user should be able to easily pass their own regularization function when initializing the solver as a callable
- no need to have such general/complex `__call__` method, it should just accept the `LinearProblem` + optional initialization

This should give us a good start to have a GCG implementation, which we can then showcase in the notebook using the DA as an example - let me know what you think!

Reply:
Thanks for the review! I agree that having an easy to use GCG solver is a good idea. 

For now, I hope to keep the more general `IterativeLinearSolver` I wrote and use it to implement an easy to use GCG.  The alternative would be to transform that solver into just a GCG solver. Let me know which direction you prefer.

Reply:
> For now, I hope to keep the more general IterativeLinearSolver I wrote and use it to implement an easy to use GCG. The alternative would be to transform that solver into just a GCG solver. Let me know which direction you prefer.

I'd prefer have it transformed into a GCG solver, we can think about the next step (more abstraction) later!